File: android/feature/profile/src/main/kotlin/com/dualmusic/feature/profile/ProfileRepository.kt
package com.dualmusic.feature.profile

import com.dualmusic.core.network.ApiClient
import com.dualmusic.core.network.Endpoint
import com.dualmusic.domain.admin.AdminEndpoints
import com.dualmusic.domain.admin.AdminUser
import com.dualmusic.domain.admin.AssignRoleRequest
import com.dualmusic.domain.admin.SettingBody
import com.dualmusic.domain.admin.SettingEnabled
import com.dualmusic.domain.auth.AuthEndpoints
import com.dualmusic.domain.auth.ChangePasswordRequest
import com.dualmusic.domain.auth.MeResponse
import com.dualmusic.domain.creator.ArtistProfile
import com.dualmusic.domain.creator.CreatorEndpoints
import com.dualmusic.domain.creator.ManagerProfile
import com.dualmusic.domain.creator.PublicProfileResponse
import com.dualmusic.domain.creator.UpdateArtistProfileRequest
import com.dualmusic.domain.creator.UpdateManagerProfileRequest
import com.dualmusic.domain.role.ApplyArtistRequest
import com.dualmusic.domain.role.ApplyManagerRequest
import com.dualmusic.domain.role.PublicSetting
import com.dualmusic.domain.role.RoleEndpoints
import com.dualmusic.domain.role.RoleRequest
import com.dualmusic.domain.user.UpdateProfileRequest
import com.dualmusic.domain.user.UserEndpoints
import com.dualmusic.domain.user.UserStats
import kotlinx.serialization.builtins.ListSerializer
import kotlinx.serialization.json.Json

/**
 * Accès REST au profil : lecture (identité + stats), **édition** (`PATCH /users/me`) et
 * **demandes de rôle** (devenir artiste/manager) avec leur gating public.
 */
class ProfileRepository(private val api: ApiClient) {

    private val json = Json { explicitNulls = false }

    /** Profil + rôles du caller. */
    suspend fun me(): MeResponse = api.request(Endpoint.get(UserEndpoints.ME), MeResponse.serializer())

    /** Statistiques agrégées (artiste/manager/fan). */
    suspend fun stats(): UserStats = api.request(Endpoint.get(UserEndpoints.MY_STATS), UserStats.serializer())

    /** Met à jour le profil (nom, bio, avatar, pays, numéro). */
    suspend fun updateProfile(request: UpdateProfileRequest) {
        val body = json.encodeToString(UpdateProfileRequest.serializer(), request)
        api.request<Unit>(Endpoint.patch(UserEndpoints.UPDATE_ME, body))
    }

    /** Programme la suppression du compte (grâce 20 jours). */
    suspend fun requestAccountDeletion() {
        api.request<Unit>(Endpoint.post(UserEndpoints.ME_DELETION))
    }

    /** Annule une suppression de compte programmée. */
    suspend fun cancelAccountDeletion() {
        api.request<Unit>(Endpoint.delete(UserEndpoints.ME_DELETION))
    }

    // --- Profils publics créateurs (artiste / manager) ---

    /**
     * Profil public artiste du caller. Lu via `GET /users/:id` (pas de `GET /artists/me`
     * côté backend) — renvoie `null` si l'utilisateur n'a pas encore de profil artiste.
     */
    suspend fun myArtistProfile(userId: String): ArtistProfile? =
        api.request(Endpoint.get(CreatorEndpoints.publicProfile(userId)), PublicProfileResponse.serializer()).artistProfile

    /** Met à jour le profil public artiste (`PATCH /artists/me`). */
    suspend fun updateArtistProfile(request: UpdateArtistProfileRequest) {
        val body = json.encodeToString(UpdateArtistProfileRequest.serializer(), request)
        api.request<Unit>(Endpoint.patch(CreatorEndpoints.ARTIST_ME, body))
    }

    /** Profil public manager du caller (`GET /managers/me`). */
    suspend fun myManagerProfile(): ManagerProfile =
        api.request(Endpoint.get(CreatorEndpoints.MANAGER_ME), ManagerProfile.serializer())

    /** Met à jour le profil public manager (`PATCH /managers/me`). */
    suspend fun updateManagerProfile(request: UpdateManagerProfileRequest) {
        val body = json.encodeToString(UpdateManagerProfileRequest.serializer(), request)
        api.request<Unit>(Endpoint.patch(CreatorEndpoints.MANAGER_ME, body))
    }

    /** Change le mot de passe (utilisateur connecté). */
    suspend fun changePassword(currentPassword: String, newPassword: String) {
        val body = json.encodeToString(
            ChangePasswordRequest.serializer(),
            ChangePasswordRequest(currentPassword = currentPassword.ifBlank { null }, newPassword = newPassword),
        )
        api.request<Unit>(Endpoint.post(AuthEndpoints.PASSWORD_CHANGE, body))
    }

    /** Candidature artiste. */
    suspend fun applyArtist(request: ApplyArtistRequest) {
        val body = json.encodeToString(ApplyArtistRequest.serializer(), request)
        api.request<Unit>(Endpoint.post(RoleEndpoints.ARTIST_APPLY, body))
    }

    /** Candidature manager. */
    suspend fun applyManager(request: ApplyManagerRequest) {
        val body = json.encodeToString(ApplyManagerRequest.serializer(), request)
        api.request<Unit>(Endpoint.post(RoleEndpoints.MANAGER_APPLY, body))
    }

    /** Mes candidatures artiste (pour connaître un éventuel statut en attente). */
    suspend fun myArtistRequests(): List<RoleRequest> =
        api.request(Endpoint.get(RoleEndpoints.ARTIST_MINE), ListSerializer(RoleRequest.serializer()))

    /** Mes candidatures manager. */
    suspend fun myManagerRequests(): List<RoleRequest> =
        api.request(Endpoint.get(RoleEndpoints.MANAGER_MINE), ListSerializer(RoleRequest.serializer()))

    /**
     * Vrai si les candidatures pour ce rôle sont ouvertes (réglage admin). Par défaut ouvert
     * si le réglage est absent/illisible.
     */
    suspend fun requestsEnabled(key: String): Boolean = settingEnabled(key, default = true)

    /** Lit un flag public `{enabled}` avec une valeur par défaut si absent/illisible. */
    suspend fun settingEnabled(key: String, default: Boolean): Boolean =
        runCatching {
            api.request(Endpoint.get(RoleEndpoints.publicSetting(key)), PublicSetting.serializer())
                .value?.enabled ?: default
        }.getOrDefault(default)

    // --- Admin ---

    /** Recherche d'utilisateurs (nom/email) — réservé admin. */
    suspend fun adminSearchUsers(query: String): List<AdminUser> =
        api.request(
            Endpoint.get(AdminEndpoints.USERS_SEARCH, mapOf("q" to query)),
            ListSerializer(AdminUser.serializer()),
        )

    /** Assigne un rôle (`fan`/`artist`/`manager`/`moderator`/`admin`) à un utilisateur. */
    suspend fun adminAssignRole(userId: String, role: String) {
        val body = json.encodeToString(AssignRoleRequest.serializer(), AssignRoleRequest(userId, role))
        api.request<Unit>(Endpoint.post(AdminEndpoints.ROLES, body))
    }

    /** Révoque un rôle d'un utilisateur. */
    suspend fun adminRevokeRole(userId: String, role: String) {
        val body = json.encodeToString(AssignRoleRequest.serializer(), AssignRoleRequest(userId, role))
        api.request<Unit>(Endpoint.delete(AdminEndpoints.ROLES, body))
    }

    /** Ouvre/ferme les candidatures d'un rôle (`manager_requests_enabled`, `artist_requests_enabled`). */
    suspend fun adminSetRequestsEnabled(key: String, enabled: Boolean) {
        val body = json.encodeToString(SettingBody.serializer(), SettingBody(SettingEnabled(enabled)))
        api.request<Unit>(Endpoint.put(AdminEndpoints.setting(key), body))
    }
}

File: ios/Packages/DualMusicKit/Sources/FeatureProfile/ProfileRepository.swift
import Foundation
import CoreNetwork
import DomainModels

/// Accès REST au profil : lecture (identité + stats), **édition** (`PATCH /users/me`),
/// **demandes de rôle** (devenir artiste/manager) avec leur gating public, et les
/// opérations **admin** (réglages plateforme + assignation de rôle).
///
/// Miroir de `ProfileRepository` côté Android.
public struct ProfileRepository: Sendable {

    private let http: HTTPClient

    public init(http: HTTPClient) {
        self.http = http
    }

    // MARK: - Profil

    /// Profil + rôles du caller.
    public func me() async throws -> MeResponse {
        try await http.request(.get(UserEndpoints.me), as: MeResponse.self)
    }

    /// Statistiques agrégées (artiste/manager/fan).
    public func stats() async throws -> UserStats {
        try await http.request(.get(UserEndpoints.myStats), as: UserStats.self)
    }

    /// Met à jour le profil (nom, bio, avatar, pays, numéro).
    public func updateProfile(_ request: UpdateProfileRequest) async throws {
        try await http.send(.patch(UserEndpoints.updateMe, body: request))
    }

    /// Programme la suppression du compte (délai de grâce de 20 jours côté backend).
    public func requestAccountDeletion() async throws {
        try await http.send(.post(UserEndpoints.meDeletion))
    }

    /// Annule une suppression de compte programmée.
    public func cancelAccountDeletion() async throws {
        try await http.send(.delete(UserEndpoints.meDeletion))
    }

    /// Change le mot de passe (utilisateur connecté).
    public func changePassword(currentPassword: String, newPassword: String) async throws {
        try await http.send(
            .post(
                AuthEndpoints.passwordChange,
                body: ChangePasswordRequest(currentPassword: currentPassword.nilIfBlank, newPassword: newPassword)
            )
        )
    }

    // MARK: - Profils publics créateurs (artiste / manager)

    /// Profil public artiste du caller. Lu via `GET /users/:id` (pas de `GET /artists/me`
    /// côté backend) — `nil` si l'utilisateur n'a pas encore de profil artiste.
    public func myArtistProfile(userId: String) async throws -> ArtistProfile? {
        try await http.request(.get(CreatorEndpoints.publicProfile(userId)), as: PublicProfileResponse.self).artistProfile
    }

    /// Profil public COMPLET d'un utilisateur (compte + profil artiste + suivi). `GET /users/:id`.
    public func publicProfile(userId: String) async throws -> PublicProfileResponse {
        try await http.request(.get(CreatorEndpoints.publicProfile(userId)), as: PublicProfileResponse.self)
    }

    /// Suit / ne suit plus un artiste. `POST` / `DELETE` `/users/:id/follow`.
    public func setFollow(userId: String, follow: Bool) async throws {
        if follow {
            try await http.send(.post(ArtistEndpoints.follow(userId)))
        } else {
            try await http.send(.delete(ArtistEndpoints.follow(userId)))
        }
    }

    /// Met à jour le profil public artiste (`PATCH /artists/me`).
    public func updateArtistProfile(_ request: UpdateArtistProfileRequest) async throws {
        try await http.send(.patch(CreatorEndpoints.artistMe, body: request))
    }

    /// Profil public manager du caller (`GET /managers/me`).
    public func myManagerProfile() async throws -> ManagerProfile {
        try await http.request(.get(CreatorEndpoints.managerMe), as: ManagerProfile.self)
    }

    /// Met à jour le profil public manager (`PATCH /managers/me`).
    public func updateManagerProfile(_ request: UpdateManagerProfileRequest) async throws {
        try await http.send(.patch(CreatorEndpoints.managerMe, body: request))
    }

    // MARK: - Candidatures de rôle

    /// Candidature artiste.
    public func applyArtist(_ request: ApplyArtistRequest) async throws {
        try await http.send(.post(RoleEndpoints.artistApply, body: request))
    }

    /// Candidature manager.
    public func applyManager(_ request: ApplyManagerRequest) async throws {
        try await http.send(.post(RoleEndpoints.managerApply, body: request))
    }

    /// Mes candidatures artiste (pour détecter un statut en attente).
    public func myArtistRequests() async throws -> [RoleRequest] {
        try await http.request(.get(RoleEndpoints.artistMine), as: [RoleRequest].self)
    }

    /// Mes candidatures manager.
    public func myManagerRequests() async throws -> [RoleRequest] {
        try await http.request(.get(RoleEndpoints.managerMine), as: [RoleRequest].self)
    }

    /// Vrai si les candidatures pour ce rôle sont ouvertes (réglage admin).
    ///
    /// Par défaut **ouvert** si le réglage est absent/illisible : on ne bloque jamais un
    /// utilisateur à cause d'un réglage manquant (même politique qu'Android).
    /// - Parameter key: `artist_requests_enabled` ou `manager_requests_enabled`.
    public func requestsEnabled(_ key: String) async -> Bool {
        await settingEnabled(key, default: true)
    }

    /// Lit un flag public `{enabled}`, avec une valeur par défaut si le réglage est absent/illisible.
    /// - Parameters:
    ///   - key: clé du réglage public.
    ///   - default: valeur de repli.
    public func settingEnabled(_ key: String, default defaultValue: Bool) async -> Bool {
        let setting = try? await http.request(
            .get(RoleEndpoints.publicSetting(key)),
            as: PublicSetting.self
        )
        return setting?.value?.enabled ?? defaultValue
    }

    // MARK: - Admin

    /// Recherche d'utilisateurs (nom/email) — réservé admin.
    public func adminSearchUsers(query: String) async throws -> [AdminUser] {
        try await http.request(.get(AdminEndpoints.usersSearch, query: ["q": query]), as: [AdminUser].self)
    }

    /// Assigne un rôle (`fan`/`artist`/`manager`/`moderator`/`admin`) à un utilisateur.
    public func adminAssignRole(userId: String, role: String) async throws {
        try await http.send(.post(AdminEndpoints.roles, body: AssignRoleRequest(userId: userId, role: role)))
    }

    /// Révoque un rôle d'un utilisateur.
    public func adminRevokeRole(userId: String, role: String) async throws {
        try await http.send(.delete(AdminEndpoints.roles, body: AssignRoleRequest(userId: userId, role: role)))
    }

    /// Ouvre/ferme les candidatures d'un rôle.
    /// - Parameters:
    ///   - key: clé de réglage plateforme.
    ///   - enabled: nouvel état.
    public func adminSetRequestsEnabled(key: String, enabled: Bool) async throws {
        try await http.send(.put(AdminEndpoints.setting(key), body: SettingBody(value: SettingEnabled(enabled: enabled))))
    }
}
